File: Services/pkg/role/service.go
package role

import (
	"context"

	"github.com/ChangerzaryX1602/SkillSync/pkg/domain"
	"github.com/ChangerzaryX1602/SkillSync/pkg/models"

	helpers "github.com/zercle/gofiber-helpers"
)

type roleService struct {
	repository domain.RoleRepository
}

func NewRoleService(repository domain.RoleRepository) domain.RoleService {
	return &roleService{repository: repository}
}

func (s *roleService) CreateRole(ctx context.Context, role models.Role) []helpers.ResponseError {
	err := s.repository.CreateRole(ctx, role)
	if err != nil {
		return []helpers.ResponseError{*err}
	}
	return nil
}

func (s *roleService) GetRole(ctx context.Context, id uint) (*models.Role, []helpers.ResponseError) {
	role, err := s.repository.GetRole(ctx, id)
	if err != nil {
		return nil, []helpers.ResponseError{*err}
	}
	return role, nil
}

func (s *roleService) GetRoles(ctx context.Context, pagination models.Pagination, search models.Search) ([]models.Role, *models.Pagination, *models.Search, []helpers.ResponseError) {
	roles, paginated, searched, err := s.repository.GetRoles(ctx, pagination, search)
	if err != nil {
		return nil, nil, nil, []helpers.ResponseError{*err}
	}
	return roles, paginated, searched, nil
}

func (s *roleService) UpdateRole(ctx context.Context, id uint, role models.Role) []helpers.ResponseError {
	err := s.repository.UpdateRole(ctx, id, role)
	if err != nil {
		return []helpers.ResponseError{*err}
	}
	return nil
}

func (s *roleService) DeleteRole(ctx context.Context, id uint) []helpers.ResponseError {
	err := s.repository.DeleteRole(ctx, id)
	if err != nil {
		return []helpers.ResponseError{*err}
	}
	return nil
}

File: hono-clean-architecture/pkg/role/service.ts
import { Role, Pagination, Search } from "../models";
import { RoleRepository, RoleService as IRoleService } from "../domain";
import { ResponseError } from "../../internal/infrastructure/custom_error";

export class RoleService implements IRoleService {
  private repository: RoleRepository;

  constructor(repository: RoleRepository) {
    this.repository = repository;
  }

  async createRole(roleData: { name: string }): Promise<ResponseError[]> {
    const { error } = await this.repository.createRole(roleData);
    if (error) {
      return [error];
    }
    return [];
  }

  async getRole(id: number): Promise<{ role: Role | null; errors: ResponseError[] }> {
    const { role, error } = await this.repository.getRole(id);
    if (error) {
      return { role: null, errors: [error] };
    }
    return { role, errors: [] };
  }

  async getRoles(
    pagination: Pagination,
    search: Search
  ): Promise<{
    roles: Role[];
    pagination: Pagination;
    search: Search;
    errors: ResponseError[];
  }> {
    const result = await this.repository.getRoles(pagination, search);
    if (result.error) {
      return { roles: [], pagination, search, errors: [result.error] };
    }
    return {
      roles: result.roles,
      pagination: result.pagination,
      search: result.search,
      errors: [],
    };
  }

  async updateRole(id: number, roleData: Partial<Role>): Promise<ResponseError[]> {
    const error = await this.repository.updateRole(id, roleData);
    if (error) {
      return [error];
    }
    return [];
  }

  async deleteRole(id: number): Promise<ResponseError[]> {
    const error = await this.repository.deleteRole(id);
    if (error) {
      return [error];
    }
    return [];
  }

  async getRoleByName(name: string): Promise<{ role: Role | null; errors: ResponseError[] }> {
    const { role, error } = await this.repository.getRoleByName(name);
    if (error) {
      return { role: null, errors: [error] };
    }
    return { role, errors: [] };
  }
}

export function newRoleService(repository: RoleRepository): IRoleService {
  return new RoleService(repository);
}
